src/io/flutter/run/daemon/FlutterApp.java

```diff
@@ -0,0 +1,233 @@
+ */
+public interface FlutterApp {
+
+  /**
+   * @return The FlutterDaemonService used to communicate with the Flutter app.
+   */
+  FlutterDaemonService getService();
+
+  /**
+   * @return The FlutterDaemonController that controls the process running the daemon.
+   */
+  FlutterDaemonController getController();
+
+  /**
+   * @return <code>true</code> if the appId has been set. This is set asynchronously after the app is launched.
+   */
+  boolean hasAppId();
+
+  /**
+   * @return The appId for the running app.
+   */
+  String appId();
+
+  /**
+   * @return The Flutter project directory.
+   */
+  String projectDirectory();
+
+  /**
+   * @return The deviceId the app is running on.
+   */
+  String deviceId();
+
+  /**
+   * @return <code>true</code> if the app can be restarted.
+   */
+  boolean isRestartable();
+
+  /**
+   * @return <code>true</code> if the app is in hot-restart mode.
+   */
+  boolean isHot();
+
+  /**
+   * @return The mode the app is running in.
+   */
+  RunMode mode();
+
+  /**
+   * @return The route parameter.
+   */
+  String route();
+
+  /**
+   * @return The target parameter.
+   */
+  String target();
+
+  /**
+   * @return The debug port used to talk to the observatory.
+   */
+  int port();
+
+  /**
+   * Stop the app.
+   */
+  void performStop();
+
+  /**
+   * Restart the app.
+   */
+  void performRestart();
+
+  /**
+   * Refresh the app.
+   */
+  void performRefresh();
+
+  /**
+   * Fetch the widget hierarchy.
+   *
+   * @return Unknown
+   */
+  Object fetchWidgetHierarchy();
+
+  /**
+   * Fetch the render tree
+   *
+   * @return Unknown
+   */
+  Object fetchRenderTree();
+}
+
+class RunningFlutterApp implements FlutterApp {
+
+  private FlutterDaemonService myService;
+  private FlutterDaemonController myController;
+  private FlutterAppManager myManager;
+  private AppStarted myApp;
+  private RunMode myMode;
+  private boolean isHot;
+  private String myRoute;
+  private String myTarget;
+  private int myPort;
+
+  public RunningFlutterApp(@NotNull FlutterDaemonService service,
+                           @NotNull FlutterDaemonController controller,
+                           @NotNull FlutterAppManager manager,
+                           @NotNull RunMode mode,
+                           boolean hot,
+                           @Nullable String target,
+                           @Nullable String route) {
+    myService = service;
+    myController = controller;
+    myManager = manager;
+    myMode = mode;
+    isHot = hot;
+    myRoute = route;
+    myTarget = target;
+  }
+
+  void setApp(AppStarted app) {
+    myApp = app;
+  }
+
+  @Override
+  public FlutterDaemonService getService() {
+    return myService;
+  }
+
+  @Override
+  public FlutterDaemonController getController() {
+    return myController;
+  }
+
+  @Override
+  public boolean hasAppId() {
+    return myApp != null && myApp.appId != null;
+  }
+
+  @Override
+  public String appId() {
```

Comment:
I'm undecided about null checks in this class. If higher-level code (like UI) is trying to use this object before the async initialization has completed then you've got bigger problems. This should be fully initialized before it gets to UI-level code so I think NPE is an appropriate action here. I could be persuaded otherwise, but some sort of exception will be thrown and if it is checked it will complicate lots of things.